File: Top_100_Liked_Questions/142._Linked_List_Cycle_II/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        # double pointer method
        if not head:
            return None
        
        slow = fast = head
        while True:
            fast = fast.next
            if not fast:
                return None
            
            fast = fast.next
            if not fast:
                return None
            
            slow = slow.next
            if slow is fast:
                break
        
        slow = head
        while slow is not fast:
            slow = slow.next
            fast = fast.next
        
        return fast

File: Top_100_Liked_Questions/142._Linked_List_Cycle_II/solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null)
            return null;

        ListNode fast = head;
        ListNode slow = head;

        while (true) {
            fast = fast.next;
            if (fast == null)
                return null;

            fast = fast.next;
            if (fast == null)
                return null;

            slow = slow.next;

            if (fast == slow)
                break;
        }

        slow = head;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }

        return fast;
    }
}